File: smoke/director_smoke.mjs
// smoke/director_smoke.mjs
// 导演台（4.4）前端产物 smoke：验证构建产物存在、入口脚本引用正确、v0 旧工程 fixture 可被迁移读取。
// 用法：node smoke/director_smoke.mjs （需先 npm run build）

import { readFileSync, existsSync, writeFileSync, mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
let failed = false;
const fail = (msg) => { failed = true; console.error('[FAIL]', msg); };
const ok = (msg) => console.log('[OK]', msg);

// 1) 构建产物
const directorHtml = join(root, 'gui/dist/director.html');
if (!existsSync(directorHtml)) {
  fail('gui/dist/director.html 不存在（请先 npm run build）');
} else {
  const html = readFileSync(directorHtml, 'utf-8');
  if (!/<script src="\.\/assets\/director-[^"]+\.js"><\/script>/.test(html)) {
    fail('director.html 未引用 director bundle 脚本');
  } else {
    ok('gui/dist/director.html 存在且引用 bundle');
  }
}

// 2) 主应用入口未被导演台构建覆盖
const mainHtml = join(root, 'gui/dist/index.html');
if (existsSync(mainHtml)) {
  const main = readFileSync(mainHtml, 'utf-8');
  if (main.includes('导演台')) {
    fail('gui/dist/index.html 被导演台构建覆盖（应为主画布入口）');
  } else {
    ok('主画布 index.html 未被覆盖');
  }
}

// 3) v0 旧工程 fixture：结构宽容、字段缺失可读（模拟 project-store.migrate 的输入契约）
const fixture = {
  format: 'icdirector',
  version: 0, // 旧版：缺失大部分可选字段
  id: 'f0000000-0000-4000-8000-0000000000f0',
  name: '旧工程 fixture',
  scene: [{ id: 'a1', kind: 'box', name: '旧盒子' }], // 无 position/rotation/scale/visible/locked
  cameras: [], // 无相机 → activeCameraId 空
  activeCameraId: '',
  references: [],
  lighting: { ambientIntensity: 1 }, // 残缺灯光 → 合并默认值
  timeline: { duration: 10, fps: 24, keyframes: [] },
  assets: [],
};
const dir = mkdtempSync(join(tmpdir(), 'director-smoke-'));
const fixturePath = join(dir, 'v0.icdirector');
writeFileSync(fixturePath, JSON.stringify(fixture, null, 2), 'utf-8');
try {
  const data = JSON.parse(readFileSync(fixturePath, 'utf-8'));
  if (data.format !== 'icdirector') fail('fixture format 校验失败');
  if (data.version !== 0) fail('fixture version 应为 0');
  // 迁移契约（与 src/director/engine/project-store.ts migrate 一致）：
  // 缺省字段应被补全为可用的默认值 —— 此处做文件级契约检查
  const migrateCompatible = {
    ...data,
    version: 1,
    scene: data.scene.map((o) => ({
      id: o.id || 'generated',
      name: o.name || '对象',
      kind: o.kind || 'box',
      position: o.position || { x: 0, y: 0, z: 0 },
      rotation: o.rotation || { x: 0, y: 0, z: 0 },
      scale: o.scale || { x: 1, y: 1, z: 1 },
      visible: o.visible !== false,
      locked: o.locked === true,
      color: o.color || '#d8d4c8',
    })),
    lighting: { ...{ ambientColor: '#f2f0ea', ambientIntensity: 0.55, keyColor: '#fff6e8', keyIntensity: 1.35, keyDirection: { x: 2.2, y: 3.4, z: 1.6 }, fillColor: '#cfe0ff', fillIntensity: 0.5, fillDirection: { x: -2.4, y: 1.2, z: -1.8 }, exposure: 1.0, background: '#1e1f24' }, ...data.lighting },
  };
  const obj = migrateCompatible.scene[0];
  if (obj.position.x !== 0 || obj.visible !== true || obj.locked !== false) {
    fail('v0 迁移缺省字段补全失败');
  } else {
    ok('v0 旧工程 fixture 迁移兼容（缺省字段补全）');
  }
} catch (e) {
  fail(`fixture 读取失败：${e.message}`);
}

if (failed) {
  console.error('DIRECTOR SMOKE: FAILED');
  process.exit(1);
}
console.log('DIRECTOR SMOKE: ALL PASS');

File: smoke/director_smoke.py
# smoke/director_smoke.py
# 导演台（4.4）最小 smoke：后端桥接契约验证 —— 创建对象 → 保存 → 重开 位置/锁定/可见性一致。
# 用法：python smoke/director_smoke.py
# 覆盖：director_save_project / director_load_project / 最近工程 / 资源校验 / 视频落盘 / 回传参数。

import json
import os
import sys
import tempfile

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from backend.director_api import DirectorAPI


class FakeSettings:
    def __init__(self):
        self.settings = {}
        self.recent = []

    def load_settings(self):
        return self.settings

    def touch_recent_project(self, path, name=''):
        self.recent.append({"path": path, "name": name})
        return {"status": "success"}

    def load_recent_projects(self):
        return {"status": "success", "projects": self.recent}

    def remove_recent_project(self, path):
        self.recent = [r for r in self.recent if r["path"] != path]
        return {"status": "success"}


class FakeAppApi:
    def __init__(self):
        self.settings = FakeSettings()

    @property
    def image(self):
        return self


def make_project(name="smoke"):
    return {
        "format": "icdirector",
        "version": 1,
        "id": "00000000-0000-4000-8000-000000000001",
        "name": name,
        "scene": [
            {
                "id": "00000000-0000-4000-8000-000000000002",
                "name": "立方体",
                "kind": "box",
                "position": {"x": 1.5, "y": 0.25, "z": -2.0},
                "rotation": {"x": 0, "y": 30, "z": 0},
                "scale": {"x": 1, "y": 1, "z": 1},
                "visible": False,
                "locked": True,
                "color": "#d8d4c8",
            }
        ],
        "cameras": [
            {
                "id": "00000000-0000-4000-8000-000000000003",
                "name": "主摄像机",
                "position": {"x": 0, "y": 2.2, "z": 6.5},
                "rotation": {"x": -12, "y": 0, "z": 0},
                "target": {"x": 0, "y": 1, "z": 0},
                "fov": 40,
                "aspect": 1.7777777777777777,
                "near": 0.1,
                "far": 1000,
                "visible": True,
                "includeInExport": True,
            }
        ],
        "activeCameraId": "00000000-0000-4000-8000-000000000003",
        "references": [],
        "lighting": {
            "ambientColor": "#f2f0ea",
            "ambientIntensity": 0.55,
            "keyColor": "#fff6e8",
            "keyIntensity": 1.35,
            "keyDirection": {"x": 2.2, "y": 3.4, "z": 1.6},
            "fillColor": "#cfe0ff",
            "fillIntensity": 0.5,
            "fillDirection": {"x": -2.4, "y": 1.2, "z": -1.8},
            "exposure": 1.0,
            "background": "#1e1f24",
        },
        "timeline": {"duration": 10, "fps": 24, "keyframes": []},
        "assets": [],
    }


def main():
    failures = []
    tmp = tempfile.mkdtemp(prefix="director_smoke_")
    api = DirectorAPI(app_api=FakeAppApi())

    # 1) 保存 → 重开：位置/锁定/可见性一致
    project_path = os.path.join(tmp, "smoke.icdirector")
    r = api.director_save_project(project_path, make_project())
    assert r["status"] == "success", r
    assert os.path.exists(project_path), "工程文件未落盘"

    r = api.director_load_project(project_path)
    assert r["status"] == "success", r
    data = r["data"]
    obj = data["scene"][0]
    assert obj["position"] == {"x": 1.5, "y": 0.25, "z": -2.0}, obj["position"]
    assert obj["locked"] is True and obj["visible"] is False, (obj["locked"], obj["visible"])
    assert data["cameras"][0]["fov"] == 40
    assert data["format"] == "icdirector"
    print("[OK] 保存→重开：位置/锁定/可见性一致")

    # 2) 最近工程
    recent = api.director_load_recent()
    assert recent["status"] == "success" and any(p["path"] == project_path for p in recent["projects"]), recent
    api.director_remove_recent(project_path)
    recent = api.director_load_recent()
    assert not any(p["path"] == project_path for p in recent["projects"])
    print("[OK] 最近工程记录 touch/load/remove")

    # 3) 资源校验（缺失文件）
    r = api.director_validate_resource(os.path.join(tmp, "not_exist.png"))
    assert r["status"] == "success" and r["exists"] is False, r
    r = api.director_validate_resource(project_path)
    assert r["status"] == "success" and r["exists"] is True, r
    print("[OK] 资源路径校验（缺失/存在）")

    # 4) 非 icdirector 文件拒绝
    bad_path = os.path.join(tmp, "bad.icdirector")
    with open(bad_path, "w", encoding="utf-8") as f:
        json.dump({"format": "icproj", "version": "3.4"}, f)
    r = api.director_load_project(bad_path)
    assert r["status"] == "error", r
    print("[OK] 拒绝非 icdirector 工程")

    # 5) 视频落盘（小 base64）
    import base64
    tiny = base64.b64encode(b"fakevideodata" * 100).decode("ascii")
    r = api.director_save_video_blob(tiny, "smoke_video.mp4")
    assert r["status"] == "success" and os.path.exists(r["path"]), r
    print("[OK] 视频落盘 director_save_video_blob")

    # 6) 回传参数脱敏转发（无主窗口时应报错但不抛异常）
    r = api.director_return_to_canvas({"kind": "png", "path": "C:/x/y.png", "projectId": "p1", "cameraId": "c1", "time": 3.5})
    assert r["status"] == "error", r  # 主窗口未设置 → 明确错误
    print("[OK] 回传通道：无主窗口时明确报错")

    # 7) GLTF 对话框不可测（GUI），但大小限制路径可测：直接调用底层不弹窗的方法不可行，跳过
    #    改为验证 ping / launch options
    api.set_launch_options({"imagePath": "C:/x.png"})
    assert api.director_get_launch_options()["imagePath"] == "C:/x.png"
    assert api.director_ping()["app"] == "icdirector"
    print("[OK] launch options / ping")

    if failures:
        print("SMOKE FAILED:", failures)
        return 1
    print("DIRECTOR SMOKE: ALL PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())
